### PR TITLE
[IMP] merge_records: be able to adjust values by sql

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
And some other minor fixes.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
You are still grouping several things with controversial ones included. Please split them for having the most obvious ones immediately merged. For the biggest one, this should be now opted-in, as existing code in OU relies on SQL merging without the values adjustment.

**Comment by MiquelRForgeFlow** [comment]
De-grouped.

> this should be now opted-in, as existing code in OU relies on SQL merging without the values adjustment.

I have done:
Selection_033

So if you don't have any field_spec but you want to adjust values, you can pass `field_spec = {}` to do the trick.

**Comment by pedrobaeza** [comment]
OK, seems a good compromise.

File: openupgradelib/openupgrade_merge_records.py [hunk 1]
```diff
@@ -612,7 +717,9 @@ def merge_records(env, model_name, record_ids, target_record_id,
         _change_foreign_key_refs(*args3)
         _change_reference_refs_sql(*args)
         _change_translations_sql(*args)
-        # TODO: Adjust values of the merged records through SQL
+        args4 = args3 + (field_spec,)
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:720]: Do this inside the `if`

File: openupgradelib/openupgrade_merge_records.py [hunk 2]
```diff
@@ -431,6 +431,111 @@ def _adjust_merged_values_orm(env, model_name, record_ids, target_record_id,
...
+            op = op or 'merge'
+            if op == 'merge':
+                first_value = _list[0]
+                if _list and not first_value:
+                    vals[column] = _list[0]
+        # Curate values that haven't changed
+        if vals[column] != _list[0]:
+            new_vals[column] = vals[column]
+    if new_vals:
+        logged_query(
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:521]: Not sure if this query should be logged, and if so, then don't double log after in debug mode. Isn't the exactly same detail?

File: openupgradelib/openupgrade_merge_records.py [hunk 3]
```diff
@@ -431,6 +431,111 @@ def _adjust_merged_values_orm(env, model_name, record_ids, target_record_id,
...
+        """, (model_table, model_name))
+    vals = {}
+    new_vals = {}
+    for column, column_type, field_type in env.cr.fetchall():
+        if (model_table, column) in exclude_columns:
+            continue
+        op = field_spec.get(column, False)
+        env.cr.execute(
+            sql.SQL(
+                """SELECT {column}
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:464]: You are doing one SELECT per column. This is not performance-wise. Read all the columns in one shot.

File: openupgradelib/openupgrade_merge_records.py [hunk 4]
```diff
@@ -431,6 +431,111 @@ def _adjust_merged_values_orm(env, model_name, record_ids, target_record_id,
...
+      to perform as values. If a field is not present here, default operation
+      will be performed.
+
+      Possible operations by field types same as _adjust_merged_values_orm.
+    """
+    if not column_exists(env.cr, model_table, 'id'):
+        return
+    env.cr.execute(
+        """
+        SELECT isc.column_name, isc.data_type, imf.ttype
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:450]: You should get fields from `ir_model_fields`, not from here, as the existing columns may not be the real fields one (you can have more and perform operations in then incorrectly). The contrary may happen the same (non stored fields, m2m/o2m fields), but you should intersect both for having the proper one, and add the TODO to handle here m2m and o2m fields.

File: openupgradelib/openupgrade_merge_records.py [hunk 5]
```diff
@@ -431,6 +431,111 @@ def _adjust_merged_values_orm(env, model_name, record_ids, target_record_id,
...
+                """SELECT {column}
+                FROM {table}
+                WHERE id IN %(record_ids)s"""
+            ).format(
+                table=sql.Identifier(model_table),
+                column=sql.Identifier(column),
+            ), {'record_ids': (target_record_id,) + tuple(record_ids)}
+        )
+        _list = [x[0] for x in env.cr.fetchall()]
+        if field_type in ('char', 'text', 'html'):
```
Comment (pedrobaeza) [inline openupgradelib/openupgrade_merge_records.py:473]: As this code is shared between both methods, I think you can extract it and pass by parameters the needed values to be handled, for saving code.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: openupgradelib/openupgrade_merge_records.py (hunk 4 above)
Comment (MiquelRForgeFlow) [reply openupgradelib/openupgrade_merge_records.py:450]: But there is a JOIN with ir_model_fields two lines below 🤔

Comment (pedrobaeza) [reply openupgradelib/openupgrade_merge_records.py:450]: OK, right, then please add the TODO for the m2m/o2m.

**Comment by MiquelRForgeFlow** [comment]
Selection_088

**Comment by MiquelRForgeFlow** [comment]
@StefanRijnhart any idea how to solve this package error?

**Comment by StefanRijnhart** [comment]
https://github.com/odoo/odoo/issues/76144 suggests setuptools<58.0. Would that be an option?

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**